pathology/hovernet/training_ignite.py

```diff
@@ -0,0 +1,303 @@
+        self.prepare_batch = PrepareBatchExtraInput(extra_keys)
+
+    def __call__(self, batchdata, device, non_blocking, **kwargs):
+        image, label, extra_label, _ = self.prepare_batch(batchdata, device, non_blocking, **kwargs)
+        all_label = {
+            HoVerNetBranch.NP: label,
+            HoVerNetBranch.NC: extra_label[0],
+            HoVerNetBranch.HV: extra_label[1],
+        }
+
+        return image, all_label
+
+
+def run(cfg):
+    log_dir = create_log_dir(cfg)
+    device = set_device(cfg)
+    # --------------------------------------------------------------------------
+    # Data Loading and Preprocessing
+    # --------------------------------------------------------------------------
+    # __________________________________________________________________________
+    # __________________________________________________________________________
+    # Build MONAI preprocessing
+    train_data, valid_data = prepare_datasets(cfg["root"])
+    train_transforms = Compose(
+        [
+            EnsureChannelFirstd(keys=("image", "label"), channel_dim=-1),
+            SplitDimd(keys="label", output_postfixes=["inst", "type"]),
+            ComputeHoVerMapsd(keys="label_inst"),
+            CastToTyped(keys=["image", "label_inst", "label_type", "hover_label_inst"], dtype=torch.float32),
+            Lambdad(keys="label", func=lambda x: x[1: 2, ...] > 0),
+            AsDiscreted(keys=["label", "label_type"], to_onehot=[2, 7]),
+            Lambdad(keys=["label", "label_inst", "label_type", "hover_label_inst"], func=lambda x: cropping_center(x, crop_shape=(164, 164))),
+            ScaleIntensityRanged( keys=["image"], a_min=0.0, a_max=255.0, b_min=0.0, b_max=1.0, clip=True),
+        ]
+    )
+    val_transforms = Compose(
+        [
+            EnsureChannelFirstd(keys=("image", "label"), channel_dim=-1),
+            SplitDimd(keys="label", output_postfixes=["inst", "type"]),
+            ComputeHoVerMapsd(keys="label_inst"),
+            CastToTyped(keys=["image", "label_inst", "label_type", "hover_label_inst"], dtype=torch.float32),
+            Lambdad(keys="label", func=lambda x: x[1: 2, ...] > 0),
+            AsDiscreted(keys=["label", "label_type"], to_onehot=[2, 7]),
+            Lambdad(keys=["label", "label_inst", "label_type", "hover_label_inst"], func=lambda x: cropping_center(x, crop_shape=(164, 164))),
+            ScaleIntensityRanged( keys=["image"], a_min=0.0, a_max=255.0, b_min=0.0, b_max=1.0, clip=True),
+        ]
+    )
+    # __________________________________________________________________________
+    # Create MONAI dataset
+    train_ds = Dataset(data=train_data, transform=train_transforms)
+    valid_ds = Dataset(data=valid_data, transform=val_transforms)
+    # __________________________________________________________________________
+    # DataLoaders
+    train_loader = DataLoader(train_ds, batch_size=cfg["batch_size"], num_workers=cfg["num_workers"], pin_memory=True)
```

Comment:
Seems you don't set `shuffle=True` for the training?

Thanks.

Reply:
Have fixed it, thx!